src/rmarkdown/draft.ts

```diff
@@ -0,0 +1,134 @@
+
+interface TemplateInfo {
+    id: string;
+    package: string;
+    name: string;
+    description: string;
+    create_dir: boolean;
+}
+
+interface TemplateItem extends QuickPickItem {
+    info: TemplateInfo;
+}
+
+async function getTemplateItems(cwd: string): Promise<TemplateItem[]> {
+    const lim = '---vsc---';
+    const rPath = await getRpath();
+    const options: cp.CommonOptions = {
+        cwd: cwd,
+        env: {
+            ...process.env,
+            VSCR_LIM: lim
+        }
+    };
+
+    const rScriptFile = extensionContext.asAbsolutePath('R/rmarkdown/templates.R');
+    const args = [
+        '--silent',
+        '--slave',
+        '--no-save',
+        '--no-restore',
+        '-f',
+        rScriptFile
+    ];
+
+    let items: TemplateItem[] = [];
+
+    try {
+        const result = await spawnAsync(rPath, args, options);
+        if (result.status !== 0) {
+            throw result.error || new Error(result.stderr);
+        }
+        const re = new RegExp(`${lim}(.*)${lim}`, 'ms');
+        const match = re.exec(result.stdout);
+        if (match.length !== 2) {
+            throw new Error('Could not parse R output.');
+        }
+        const json = match[1];
+        const templates = <TemplateInfo[]>JSON.parse(json) || [];
+        items = templates.map((x) => {
+            return {
+                alwaysShow: false,
+                description: `{${x.package}}`,
+                label: x.name,
+                detail: x.description,
+                picked: false,
+                info: x
+            };
+        });
+    } catch (e) {
+        console.log(e);
+        void window.showErrorMessage((<{ message: string }>e).message);
+    }
+
+    return items;
+}
+
+async function launchTemplatePicker(cwd: string): Promise<TemplateItem> {
+    const options: QuickPickOptions = {
+        matchOnDescription: true,
+        matchOnDetail: true,
+        canPickMany: false,
+        ignoreFocusOut: false,
+        placeHolder: '',
+        onDidSelectItem: undefined
+    };
+
+    const items = await getTemplateItems(cwd);
+
+    const selection: TemplateItem = await window.showQuickPick<TemplateItem>(items, options);
+    return selection;
+}
+
+async function makeDraft(file: string, template: TemplateItem, cwd: string): Promise<string> {
+    const fileString = ToRStringLiteral(file, '');
+    const cmd = `cat(normalizePath(rmarkdown::draft(file='${fileString}', template='${template.info.id}', package='${template.info.package}', edit=FALSE)))`;
+    return await executeRCommand(cmd, cwd, (e: Error) => {
+        void window.showErrorMessage(e.message);
+        return '';
+    });
+}
+
+export async function newDraft(): Promise<void> {
+    const cwd = getCurrentWorkspaceFolder()?.uri.fsPath ?? os.homedir();
+    const template = await launchTemplatePicker(cwd);
+    if (!template) {
+        return;
+    }
+
+    if (template.info.create_dir) {
+        const uri = await window.showSaveDialog({
```

Comment:
The `.Rmd` extension might be a bit confusing here. I did not find a really clean solution to configure the SaveDialog to specify a folder, but the following could be used:
``` ts 
        let defaultPath = path.join(cwd, 'draft');
        let i = 1;
        while(fs.existsSync(defaultPath)){
            defaultPath = path.join(cwd, `draft_${++i}`);
        }

        const uri = await window.showSaveDialog({
            defaultUri: Uri.file(defaultPath),
            filters: {
                'Folder': ['']
            },
            saveLabel: 'Create Folder',
            title: 'R Markdown: New Draft'
        });
```

The open dialogue seems to have an option to select only folders, but does not allow to select non-existing folders

Reply:
Yes, it is something I'm also considering. I tried `showOpenDialog` with folder-only [options](https://code.visualstudio.com/api/references/vscode-api#OpenDialogOptions), but it didn't work smoothly for saving purpose.

Reply:
Also, `rmarkdown::draft()` will throw error if the target file or folder exists, which I have not handled yet.

Reply:
`rmarkdown::draft()` will create a folder with last file extension removed. For example, `draft.Rmd` will become `draft`; `draft.test.Rmd` will become `draft.test`.

I update the code to check if this folder already exists and let user confirm whether to remove it before creating the draft folder.